darts/dataprocessing/pipeline.py

```diff
@@ -0,0 +1,178 @@
+"""
+Pipeline
+--------
+"""
+from copy import deepcopy
+from typing import Sequence, Union, Iterator
+
+from darts.logging import raise_if_not, get_logger
+from darts import TimeSeries
+from darts.dataprocessing.transformers import BaseDataTransformer, InvertibleDataTransformer, FittableDataTransformer
+
+logger = get_logger(__name__)
+
+
+class Pipeline:
+    def __init__(self,
+                 transformers: Sequence[BaseDataTransformer[TimeSeries]],
+                 copy: bool = False):
+        """
+        Pipeline combines multiple data transformers chaining them together.
+
+        Parameters
+        ----------
+        transformers
+            Sequence of data transformers.
+        copy
+            If set makes a (deep) copy of each data transformer before adding them to the pipeline
+        """
+        raise_if_not(all((isinstance(t, BaseDataTransformer)) for t in transformers),
+                     "transformers should be objects deriving from BaseDataTransformer", logger)
+
+        if transformers is None or len(transformers) == 0:
+            logger.warning("Empty pipeline created")
+            self._transformers: Sequence[BaseDataTransformer[TimeSeries]] = []
+        elif copy:
+            self._transformers = deepcopy(transformers)
+        else:
+            self._transformers = transformers
+
+        self._invertible = all((isinstance(t, InvertibleDataTransformer) for t in self._transformers))
+
+    def fit(self, data: TimeSeries):
+        """
+        Fit all fittable transformers in pipeline.
+
+        Parameters
+        ----------
+        data
+            TimeSeries to fit on.
+        """
+
+        # Find the last fittable transformer index
+        # No need to fit (and thus transform) after this index, possibly saving a fair bit of time
+        last_fittable_idx = -1
+        for idx, transformer in enumerate(self._transformers):
```

Comment:
Wouldn't it be enough to simplify and just have something like:

```
for transformer in self._transformers:
    if isinstance(transformer, FittableDataTransformer):
        transformer.fit(data)
    data = transformer.transform(data)
```

Reply:
Well I thought about this, and decided to go with the longer version to improve performance, but it might not be worth it?

The problem with the solution you propose (which is actually the same as calling `self.fit_transform(data)` and discarding the result) is that it runs at least one superfluous transform (the last one), and possibly more.

If for example the only the first transformer in the pipeline is fittable, there is no need to transform the data at all to fit other transformers further in the pipeline (because there are none). I'm not sure exactly how costly a transformation can be, but I thought it'd be nicer to avoid those unnecessary transforms. What do you think ?

Reply:
OK, fair enough, agree :)